apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts

```diff
@@ -834,12 +883,16 @@ export class TWRPortfolioCalculator extends PortfolioCalculator {
           investmentValuesAccumulatedAtStartDateWithCurrencyEffect
         );
 
-      const dates = eachDayOfInterval({
-        end: endDate,
-        start: startDate
-      }).map((date) => {
-        return format(date, DATE_FORMAT);
-      });
+      let dates = this.cachedEachDayOfInterval[dateRange];
+      if (dates === undefined) {
+        dates = eachDayOfInterval({
+          end: endDate,
+          start: startDate
+        }).map((date) => {
+          return format(date, DATE_FORMAT);
+        });
+        this.cachedEachDayOfInterval[dateRange] = dates;
+      }
 
       let average = new Big(0);
       let dayCount = 0;
```

Comment:
That's an interesting finding! 👍🏻

There might even be a superior optimization by just replacing

```
for (const date of dates) {
```
with
```
for (const date of Object.keys(chartDateMap)) {
```
below (and remove `const dates = ...` completely).

Reply:
I took a quick look at doing that, and I think I have some questions. I don't fully grasp everything that the portfolio calculator is doing, so I might be missing some bigger picture context here, but wouldn't that proposed change result in different values in the end?

For instance, when the outer loop is at `dateRage = '1d'`, the size of `dates` will be 2, right? But if we use `for (const date of Object.keys(chartDateMap) { ... }`, then it will go through way more days. Wouldn't this significantly affect the resulting values?

One other option here that I just thought of is to do something like the pseudocode below instead of creating/caching the dates:
```
currentDate = startDate;
while (currentDate < endDate) {
  ...
  currentDate += 1;
}
```

Reply:
Which outer loop do you mean? I don't think the date range matters here. The portfolio calculator always calculates the snapshot for `max` while `chartDateMap` includes all relevant dates.

Reply:
I meant this loop:
```
for (const dateRange of ['1d', '1y', '5y', 'max', ...]) {
  ...
  const dates = eachDayOfInterval(...);
  ...
  for (const date of dates) {
    ...
  }
}
```


> I would expect that the helper function eachDayOfInterval is already doing something very similar. No?

Yeah, but I think the slowdown, at least in my case, is due to `eachDayOfInterval` allocating big arrays for the `'5y'` and `'max'` (11 years for me) ranges. Manually incrementing the days would avoid that overhead.

Reply:
> Wouldn't this significantly affect the resulting values?

I agree! So your caching approach is better. However, we could additionally filter the result of `eachDayOfInterval()` by checking if the date is relevant (part of `chartDateMap`) to further optimize. What do you think?

Reply:
Here's an example to show what I think I'm failing to understand. Let's say we have these dates and values in `investmentValuesAccumulatedWithCurrencyEffect`:
2024-01-01: 0
2024-01-02: 0
2024-01-03: 0
2024-01-04: 1
2024-01-05: 1

The `for (const date of dates)` loop computes the average for these days, which would be `0.4` considering all dates.

Now if we filter by the dates present in `chartDateMap`, and let's say `chartDateMap` only has 2 of those dates, the first and the last. Then the average would result in `0.5`.

Would that be ok?

Reply:
> Now if we filter by the dates present in chartDateMap, and let's say chartDateMap only has 2 of those dates, the first and the last. Then the average would result in 0.5.

`chartDateMap` includes all dates where you have an activity. So you will get the dates where the investment value actually changes. Not quite sure if it makes a difference, though.

Reply:
Ah! Since `getChartDateMap` uses `MAX_CHART_ITEMS` I mistakenly assumed it would contain at most `MAX_CHART_ITEMS`. I'll try using `chartDateMap` here and report back.

Reply:
Using `chartDateMap`, this actually reduced the time spent per symbol on this `dateRange` loop by an extra 5ms. So now, on my machine, it's at 9.7ms instead of 100ms.